### PR TITLE
Indicator when a candy upgrade is available

### PR DESCRIPTION
**Pull request by josericardo-fo**
Adds a candy indicator when a passive/cost reduction upgrade can be purchased.
Also addds the setting Candy Upgrade Icon with three options ('Off' by default):

image

'No' doesn't show icons, 'Passives Only' only shows icons when you have enough candies to unlock that Pokémon's Passive Ability and 'On' shows icons when you have enough candies to unlock its Passive or Reduce its cost.

Examples: 'Off' and everything has enough candies

image

'Passives Only' and the last row had their passives unlocked (since Snorlax has Munchlax as its baby, the icon still appears if you can unlock its baby Passive)

image

'Passives Only', the last row had their passives unlocked, the first row has their costs reduced twice and the second row had their costs reduced once and the third row has everything unlocked

image

The icons also update themselves when you buy something

[Purchasing Tests](https://streamable.com/465w00)

### CODE REVIEW

File: src/ui/starter-select-ui-handler.ts [hunk 1]
```diff
@@ -960,6 +984,7 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
                       if (!success)
                         return this.scene.reset(true);
                     });
+                    if (this.scene.candyUpgradeIconsMode !== 0) {this.updateCandyUpgradeIcon(this.cursor)} // If the setting is not set to 0, update the candy upgrade icon
```
Comment (CodeTappert) [inline src/ui/starter-select-ui-handler.ts:987]: Add comment ABOVE the line not next to it

Also the code that should be executed if the condition is true should be in the line after the condition
```

if (this.scene.candyUpgradeIconsMode !== 0)
 // If the setting is not set to 0, update the candy upgrade icon
 {this.updateCandyUpgradeIcon(this.cursor)}
```

if it doesnt work like that then do

```

if (this.scene.candyUpgradeIconsMode !== 0) {
 // If the setting is not set to 0, update the candy upgrade icon
 {this.updateCandyUpgradeIcon(this.cursor)}
}
```

File: src/ui/starter-select-ui-handler.ts [hunk 2]
```diff
@@ -984,6 +1009,7 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
                       if (!success)
                         return this.scene.reset(true);
                     });
+                    if (this.scene.candyUpgradeIconsMode === 2) {this.updateCandyUpgradeIcon(this.cursor)} // If the setting is set to 2, update the candy upgrade icon
```
Comment (CodeTappert) [inline src/ui/starter-select-ui-handler.ts:1012]: Same as https://github.com/pagefaultgames/pokerogue/pull/1083/files#r1605887572

**Comment by josericardo-fo** [comment]
Finished. Waiting for review.

File: src/ui/starter-select-ui-handler.ts [hunk 3]
```diff
@@ -1758,6 +1819,31 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
     this.starterValueLabels[cursor].setShadowColor(this.getTextColor(textStyle, true));
   }
 
+  updateCandyUpgradeIcon(cursor: integer): void {
+    const speciesId = this.genSpecies[this.getGenCursorWithScroll()][cursor].speciesId;
+
+    switch (this.scene.candyUpgradeIconsMode) {
+      case 0: // 'Off' mode
+          return
```
Comment (CodeTappert) [inline src/ui/starter-select-ui-handler.ts:1827]: mising ;

File: src/ui/starter-select-ui-handler.ts [hunk 4]
```diff
@@ -984,6 +1012,10 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
                       if (!success)
                         return this.scene.reset(true);
                     });
+                    // If the setting is set to 2, update the candy upgrade icon
+                    if (this.scene.candyUpgradeIconsMode === 2) {
+                      this.updateCandyUpgradeIcon(this.cursor)
```
Comment (CodeTappert) [inline src/ui/starter-select-ui-handler.ts:1017]: missing ;

File: src/ui/starter-select-ui-handler.ts [hunk 5]
```diff
@@ -960,6 +984,10 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
                       if (!success)
                         return this.scene.reset(true);
                     });
+                    // If the setting is not set to 0, update the candy upgrade icon
+                    if (this.scene.candyUpgradeIconsMode !== 0) {
+                      this.updateCandyUpgradeIcon(this.cursor)
```
Comment (CodeTappert) [inline src/ui/starter-select-ui-handler.ts:989]: missing ;

**Comment by j-diefenbach** [comment]
LGTM although the one line comments might be a little excessive. E.g. you could change 'on mode' 'off mode' etc. to a single line comment above or further above clarifying which mode is what, a decent amount of that code is pretty clear in what it does so comments aren't necessary for those cases imo.

**Comment by bennybroseph** [comment]
https://github.com/pagefaultgames/pokerogue/assets/13838608/533561c2-6cc4-4d77-8e17-60697bfd8ca5

https://github.com/pagefaultgames/pokerogue/assets/13838608/b6770c38-771e-46f6-9fa8-f4ab903fbbf7

https://github.com/pagefaultgames/pokerogue/assets/13838608/1d2873d3-bcc6-4303-bddb-e47406a5102b

https://github.com/pagefaultgames/pokerogue/assets/13838608/40fd2c51-bf5d-448a-b514-3dd0edb9982d

**Comment by josericardo-fo** [comment]
After changing settings, some starters Y position are wrong.

image

**Comment by josericardo-fo** [comment]
Ready for review

File: src/ui/starter-select-ui-handler.ts [hunk 6]
```diff
@@ -706,6 +750,165 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+      || species?.getRootSpeciesId(false) !== species?.speciesId) { // Pokemon is not the base evolution and can't use candy
+      // Set all icons as hidden and exit early
+      this.candyUpgradeIcon[index].setVisible(false);
+      this.candyUpgradeOverlayIcon[index].setVisible(false);
+
+      return;
+    }
+
+    if (!slotVisible) {
+      console.log("Yeah");
```
[the remark this answers is not in the document]
Comment (josericardo-fo) [reply src/ui/starter-select-ui-handler.ts]: fixed

**Review by molker — APPROVED** [review]

**Pull request merged**